src/core/thread/mesh_forwarder.cpp

```diff
@@ -754,163 +754,149 @@ otError MeshForwarder::UpdateIp6Route(Message &aMessage)
 
     aMessage.Read(0, sizeof(ip6Header), &ip6Header);
 
-    switch (netif.GetMle().GetRole())
+    VerifyOrExit(!ip6Header.GetSource().IsMulticast(), error = OT_ERROR_DROP);
+
+    // 1. Choose correct MAC Source Address.
+    GetMacSourceAddress(ip6Header.GetSource(), mMacSource);
+
+    // 2. Choose correct MAC Destination Address.
+    if (netif.GetMle().GetRole() < OT_DEVICE_ROLE_CHILD)
     {
-    case OT_DEVICE_ROLE_DISABLED:
-    case OT_DEVICE_ROLE_DETACHED:
+        // Allow only for Link-local unicasts and multicasts.
         if (ip6Header.GetDestination().IsLinkLocal() || ip6Header.GetDestination().IsLinkLocalMulticast())
         {
             GetMacDestinationAddress(ip6Header.GetDestination(), mMacDest);
-            GetMacSourceAddress(ip6Header.GetSource(), mMacSource);
         }
         else
         {
             ExitNow(error = OT_ERROR_DROP);
         }
 
-        break;
+        ExitNow();
```

Comment:
Is it clearer to merge them into one as @librasungirl suggests? I think that indicates `if (netif.GetMle().GetRole() < OT_DEVICE_ROLE_CHILD`, the processing should end in this if body.